### ISSUE TITLE
`build-tool` means multiple things

### ISSUE THREAD

**Issue by ekmett** [commercialhaskell/stack#3385]
I have a package which provides an executable that is required to be built by elements of the test suite. The mechanism provided by cabal do to this is to name a `build-tool` that is the name of the executable that will be produced by the local package. This then causes invocations of `cabal test` in a clean directory to first ensure that the executable is built before running the test.

When building with `cabal` this works fine, but when I go to invoke `stack` I get a warning:

> No packages found in snapshot which provide a "coda" executable, which is a build-tool dependency of "coda"

According to the documentation for [`build-tool`](http://cabal.readthedocs.io/en/latest/developing-packages.html#pkg-field-build-tools) the first thing checked is for an executable in the same package, then it falls back on finding other packages. `stack` seems to skip immediately to step 2.

**Comment by snoyberg** [comment]
This is newly added documentation AFAICT, and was not present in previous versions of the Cabal documentation. Due to the lack of a clear definition, most tooling that I've interacted with (including discussions with the Nix maintainers) have had to evolve custom meanings. For example, in the page you link, it says:

> One of a hard-coded set of packages containing common build tools

We have explicitly _not_ followed such a hard coding in Stack, instead creating a tool map from the snapshot in use. We've also futzed around a bit with the meaning of this field, such as whether it refers to executable names or package names (I can't find the issue number that discussed this right now).

All that said: the newest version of Stack on master does incidentally solve your specific case, since it includes local packages in the creation of a tool map, and therefore identifies that `coda`-the-build-tool is already present in the install plan, bypassing the warning.

So, tl;dr:

* It's likely too late to change the assumed meaning of the `build-tools` field, despite updates to the definition in the Cabal documentation, since the rest of the world already assumed a different meaning
* But your acute issue shouldn't occur in the next release

I'm not terribly opposed to the idea of adding a special case

**Comment by ekmett** [comment]
I'll take an accidental fix. ;)

I don't particularly care what reasoning wins so long as I have the ability to actually make a test that depends on an executable that works on both `cabal install` based builds and `stack install` builds.

My understanding is cabal side they are moving away from this field because they have similar distaste for its muddled semantics with a desire to shift to a `tool-depends` field that names `package:executable` pairs for tooling.

**Comment by snoyberg** [comment]
That's my understanding too; the `tool-depends` field looks like a significant improvement. Since this is already addressed on master, closing.
